_posts/2025-05-16-Introduction_to_bidirectional-languages-in-your-website.md

```diff
@@ -0,0 +1,127 @@
+
+Let's see what happens when I change the language without doing anything with RTL layout in mind.
+
+![[simple website 2.png]](/img/2025-05-16-introduction-to-bidirectional-languages-in-your-website/2.png)
+
+Not as bad as you would think but there are few things that are wrong here. The header should be flipped and the names `Ninni` and `Eino` beside the pictures should be aligned to the left side.
+
+Fortunately there is a quick fix for this. We can set the directionality for the element's using the `dir` attribute. We can set it on specific elements but for this case we want it on the whole html.
+
+```html
+<html dir="rtl"></html>
+```
+
+Let's change the directionality every time the language changes.
+
+```typescript
+useEffect(() => {
+  const dir = i18n.dir(i18n.language);
+  document.documentElement.dir = dir;
+}, [i18n, i18n.language]);
+```
+
+Let's see what's changed.
+
+![[simple website 3.png]](/img/2025-05-16-introduction-to-bidirectional-languages-in-your-website/3.png)
+
+Everything is now flipped correctly and we are finished here right? Well there seems to be few things still that are wrong here.
+
+### Margin and padding
+
+If we take a closer look we can see that the margins and paddings aren't flipped when the directionality changes.
+
+There are out-of-the-box solutions that deals with margins and padding etc. automatically such as [rtlcss](https://rtlcss.com/learn/) and [tailwindcss-rtl](https://www.npmjs.com/package/tailwindcss-rtl). For demonstration purposes I will show you how to deal with them manually.
+
+When creating a website with support for both directionalities you should avoid using `margin` and `padding`. Instead we should use `margin-inline-start`, `margin-inline-end`,
+`padding-inline-start` and `padding-inline-start` which aligns items depending on the directionality.
+
+[Tailwind](https://tailwindcss.com/blog/tailwindcss-v3) offers RTL and LTR modifiers which can be used.
+
+```html
+<header className="... ltr:pl-4 rtl:pr-4"></header>
+```
+
+![[simple website 4.png]](/img/2025-05-16-introduction-to-bidirectional-languages-in-your-website/4.png)
+
+Now the margins are fixed on the header. There is still one small quirk when setting directionality for the page.
+
+I didn't translate this text just to demonstrate this special case when having LTR text in a RTL directional page.
+
+![[simple website 5.png]](/img/2025-05-16-introduction-to-bidirectional-languages-in-your-website/5.png)
+
+### Neutral characters
+
+Some Unicode characters are not associated with some directionality e.g. numbers and punctuation. These characters are considered to be weak or neutral. The [Unicode bidirectional algorithm](https://www.w3.org/International/articles/inline-bidi-markup/uba-basics) usually handles these situations automatically by looking the characters around the neutral or weak character but not always.
```

Comment:
"Some Unicode characters are not associated with some directionality e.g. numbers and punctuation." -> Maybe better sentence structure if you write this something like "Some Unice characters such as numbers and punctuation are not associated with any directionality."

Reply:
by looking the characters -> by looking at the characters